### ISSUE TITLE
Making the Xournal++ a better open source alternative to OneNote

### ISSUE THREAD

**Issue by thezeroth** [xournalpp/xournalpp#2000]
Its very nice to see that my favorite old open source note taking software finally getting some love, I always thought this software has the potential to be a better alternative to OneNote.  So I am creating this Issue to gather ideas and suggestions on what features needs to be added, what changes to user interface is needed to beat OneNote.

### My Initial suggestions
**1) Making Xournal touch friendly**
when this software was initially developed most devices with pen lacked touch input, as a result currently it doesn't utilize touch well
**2) Better Organization of larger Note Journals**
currently we can only add new pages, OneNote has tabs on top of the page so one can organize their notes into different topics, each topic can have many section tap on the right side, then each section can have many pages.
**3) Ability to sync Notes between devices**
Currently we can use open source PtP syncing software like Syncthing to Sync note files securely even if the devices are not on one network. So not much needs to be done to add sync ability, we have to consider what happens if someone edits the same note on both devices without syncing them. and how to deal with it if we create a standard directory structure to achieve my second suggestion above.

**Comment by LittleHuba** [comment]
You are most welcome to create some mocks for a more touch friendly GUI. Including what gestures you'd like to have supported and so on. That should cover the first two points. 

The third point is already in work. We are working on a file format supporting collaborative editing as we speak. But that will take quite some time as it is no easy thing to get right.

**Comment by daljit97** [comment]
@LittleHuba the problem I see with a zip based file format is that each time an Xournal file is updated the entire file has to be reuploaded to the cloud (e.g Onedrive in Windows 10). This could be solved by allowing to save files in folders instead so only changes to the modified files would be synced.

**Comment by thezeroth** [comment]
@daljit97 I agree, I think best approach is to get Xournal to manage a sub-directory for each notebook:

Notebook1 (root folder)
-Project1 (folder)
--Milestone1 (subfolder)
---Task1.ext (file)
---Task2.ext (file)
--Milestone2 (subfolder)
---Task1.ext (file)
---Task2.ext (file)
-Project2 (folder)
--Milestone1 (subfolder)
---Task1.ext (file)
---Task2.ext (file)
--Milestone2 (subfolder)
---Task1.ext (file)
---Task2.ext (file)
-Research (folder)
--External (subfolder)
---Task1.ext (file)
---Task2.ext (file)
--Internal (subfolder)
---Task1.ext (file)
---Task2.ext (file)

Say we point Xournal to "Notebook1" as the root directory, Xournal will show 3 tabs on the top (Project1, Project2 and Research), then When a tab is chosen, it will show tabs on the right side.
 _____________
- | **Project1** | - - | Project2 | - - | Research | -
................................................................................................................................................ |-**Milestone1**-|
................................................................................................................................................ |--Task1------|
................................................................................................................................................ |--Task2------|
................................................................................................................................................ |-Milestone2-|
................................................................................................................................................ |--Task1------|
................................................................................................................................................ |--Task2------|

Say we point Xournal to "Research" as the root directory, Xournal will show 2 tabs on the top (External, Internal), then When a tab is chosen, it will show tabs on the right side.
 _____________
- | **External** | - - |Internal | - 
................................................................................................................................................ |-**Task1**-------|
................................................................................................................................................ |-Task2-------|

Advantages:

1. Part of a notebook can be easily shared, just send them the sub-directory.
2. When syncing, only edited pages need to be transferred.
3. Part of the notebook can be encrypted
4. Permission to parts of notebook can be easily set.
5. you can easily combine several notebooks into one.
6. Syncthing can sync notebooks down to page level.
7. each file can be individually compressed for less bandwidth or not (in case notebook used only inside internal network)

**Comment by LittleHuba** [comment]
You understood the approach wrong. Zip files are not about many documents in one zip but about one document with multiple pages in one zip. Should you want to separate your document into many files that is up to you but people should have the ability to put many pages in one file. 

Additionally what you did not think about is that file permissions are a problematic thing when sharing directories. You mess them up as soon as you pack the files into some kind of archive.

**Comment by LittleHuba** [comment]
Another thing your approach does not consider is the big overhead in time and power consumption you get for disk access. With breaking documents down to pages you will have tremendous disk activity for documents when it is unnecessary in most cases. 

What you propose is still a good idea but one level higher than page level please.

**Comment by daljit97** [comment]
@LittleHuba yes, I agree on that. But was suggesting was the possibility of storing documents into folders with separate files. For example, each document would be a folder that contains three folders: "Images", "Ink", "Text" (maybe additional categories as well). Inside the main folder you'll have a main file that can be used to open Xournal. Then inside the "Ink" folder, the ink will be stored in binary files such that each file has a maximum limit of around 3Mb and if the strokes inside the document are modified then only the corresponding ink file that contains those would need to be modified. Obviously this approach has many downsides, as you mentioned increased disk usage would be one of them, but I think this can be limited if the files are appropriately sized. Also, folders are obviously a less elegant solution than single files. On the other hand, this would allow a significant reduction in bandwidth for people (I think nearly everyone backs up their notes somehow) who use third party services to sync their files into the cloud.

**Comment by LittleHuba** [comment]
For collaboration there is no benefit as we work with deltas. So only data that was changed is actually transferred between users. 

One other issue is that your solution would invite people to tamper directly with the internal data. That is dangerous and could lead to corrupted files. We do have internal state that should not be confused by doing stuff that should not be done. 

Also backup solutions migrating documents non-atomically could corrupt the document itself. We do have dependencies between files within the zip. That would open another Box of the Pandora. Only think about multiple people working on one document uploading it to the same server. It would most certainly end in some mixed state with backup conflicts all over the place. 

If you fear about bandwidth, I'd recommend you searching for a backup solution that uses deltas.

**Comment by thezeroth** [comment]
@LittleHuba I should clarify, I never meant one page, one file. I am suggesting one file for each section of the notebook. each section should be able to have as many pages as possible.  for example:

Notebook1 (root folder)
-Project1 (folder)
--Milestone1 (subfolder)
---Task1.ext (file)

Task1.ext will contain all the pages in Task1 section.

Regarding permission, I agree with you that "file permissions are a problematic thing" specially because Xournal as a multi platform app, would run on many different file systems. I suggest to get ahead of OneNote and anything else out there and use a Public and Private Key System to manage access.  here is an example of  Public and Private Key implementation:

On first use Xournal will generate a pair of public and private keys for the owner (or owner can import their own key pair). then everything they enter into the notebook is signed by their private key.  and everyone else can verify that with owner's public key. owner can give access to collaborators by signing collaborators public key with his owner private key. everything a collaborator is entering is signed with collaborator private key and can be verified by everyone else using collaborator's public key.

To make a part of notebook private, Xournal can encrypt it with user's private key

**Comment by LittleHuba** [comment]
So what you in essence want is a high level format as I recommended earlier on. You just use different names (sections instead of documents). I'd introduce a second file type for this as there are some implications that make those notebooks too different from other filetypes to keep good import export compatibility with other file formats. 

The use of key pairs for access rights is interesting. We will not be able to implement this on an element basis but it would be possible on document basis. The computation effort would be to much if we would decrypt and encrypt elements individually. One "problem" I see is that we won't be able to strongly couple identities to keys. Without a central authority the keys are just keys. We could go and use something like GnuPG. But it has no real advantage for small groups of users where no big user base is available to ensure trust. But maybe fixing trust issues is just out of scope... 

@lehmanju what do you think about this?

**Comment by LittleHuba** [comment]
Oh and about the tabs of OneNote: I hated those. They were thought to introduce some kind of separation into notebooks but they fall short of being a good tree-like hierarchy where you can actually store documents like you want to and not just in some awful list of ugliness. 😉

**Comment by lehmanju** [comment]
Implementing access control for a P2P application is too much effort and makes little sense in my opinion. Look at IPFS or any other P2P system, access control is done via sharing only accessible data. If you share all of the data and only make some part available by encrypting it, it is actually easier to split your data into multiple documents and only share the document with the data that you want to share. Once you have given access you can not revoke it and you naturally can not prevent somebody else from forwarding it. In our application we could however introduce multiple tabs for different documents. This way you could edit multiple files at once while we keep documents separate. With our modular graphical interface this isn't that hard to achieve and would spare us from encryption specifics.

Internally, data will be encrypted though. Before a document is sent over the network it will be encrypted by the CRDT library, but this is just transport encryption.

**Comment by LittleHuba** [comment]
> @thezeroth: each section should be able to have as many pages as possible.

In that regard, it might be interesting to introduce chapters as a level inbetween documents and their pages. See the BNF below:

```
document = chapter+ .
chapter = chapter+ | page+ .
page = ...
```

> @lehmanju: Implementing access control for a P2P application is too much effort and makes little sense in my opinion.

Two solutions to circumvent all ths AC stuff:
1. We could introduce Elements that link other documents. When you click on one such element, the linked document is openend in a new tab within your instance of Xournal++. 
2. One could allow chapters to "reference" another document. Xournal++ would open these referenced documents seamlessly as if the pages (and chapters) of the document would be part of the main document itself. The documents could be simply handled as attachments to the document itself, allowing for internal and external storage to simplify sharing.

**Comment by tdyckerh** [comment]
@LittleHuba: Is there an update on the status of the collaborative editing? I've tried
       https://github.com/lovasoa/whitebophir
which works quite well for me, but it would be awesome to have all the features of xournalpp on top.

**Comment by LittleHuba** [comment]
Unfortunately no. I don't really have the time anymore to work actively on the project so I have to restrict myself to checking PRs and some background work. Thus this issue is currently open for adoption. I'm sorry for the bad news.

**Comment by LittleHuba** [comment]
@tdyckerh I answered your mail but got some awkward error from my mailserver that it could not deliver my answer. Fell free to reach out to me with your private mail or via Gitter, should the mail not have arrived.

**Comment by rolandlo** [comment]
As explained in #799 Xournal++ notes can be conveniently organized into notebooks in [Joplin](https://joplinapp.org/). So I am not so sure, we should implement it in Xournal++.

**Comment by profiluefter** [comment]
Maybe there could be a file system view similar to how it works in Notepad++: (Sorry for the German UI)
![notepad++_1xxUV4Majn](https://user-images.githubusercontent.com/18028031/128606408-3dfbb13f-55f2-498f-a6b7-bb110ac42219.png)
So you would select a root folder and the contents of that folder would be shown in a tree-like structure in a side bar. Then you could have many individual journal files in there and structure them using folders. I think this approach would be rather easy to implement and not require a major rework or a change in the file format. Of course the drawback with much disk i/o still applies if you change the journal often but there would be no problem with synchronizing/backing up individual files as they are all self-contained. 
In the future it could then be extended so that the structure is not based on the file system but rather links to individual files but for me using the file system would be the perfect solution!

**Comment by rafisics** [comment]
Any update on syncing notes on different devices?

**Comment by profiluefter** [comment]
@rafisics I'd strongly recommend using Syncthing for file synchronization

**Comment by flckf** [comment]
> 
Hi @profiluefter have you made any progress on your fork? I'm also very keen to have something like this!

**Comment by profiluefter** [comment]
> Hi @profiluefter have you made any progress on your fork? I'm also very keen to have something like this!

Sadly not really, didn't have time for it in the end and I started just using bigger files. I do still think that the idea has potential.

As far as I remember, the core functionality already sort of worked on my fork but was missing a lot of polishing and the ability to select a folder.

I might come back to it after I finished writing my thesis but for now I'm not working on it. It's not as complicated as one might think so if someone wants to continue with the implementation, feel free to do that!
